tests/cases/compiler/isArray.ts

```diff
@@ -1,9 +1,86 @@
-var maybeArray: number | number[];
+// @strictNullChecks: true
 
+interface MyArray<T> extends Array<T> { manifest: any; }
+interface MyReadOnlyArray<T> extends ReadonlyArray<T> { manifest: any; }
 
-if (Array.isArray(maybeArray)) {
-    maybeArray.length; // OK
+function fn1(arg: string | string[]) {
+    if (Array.isArray(arg)) arg.push(""); // Should OK
+}
+
+function fn2(arg: unknown) {
+    if (Array.isArray(arg)) arg.push(""); // Should OK
+}
+
+function fn3(arg: object) {
+    if (Array.isArray(arg)) arg.push(""); // Should OK
+}
+
+function fn4(arg: {}) {
+    if (Array.isArray(arg)) arg.push(""); // Should OK
+}
+
+function fn5(arg: string | ReadonlyArray<string>) {
+    if (Array.isArray(arg)) arg.push(10); // Should FAIL
+    if (Array.isArray(arg)) arg.push(""); // Should FAIL
+    if (Array.isArray(arg)) arg.indexOf(""); // Should OK
+    if (!Array.isArray(arg)) arg.toUpperCase(); // Should OK
+}
+
+function fn6(arg: string | string[]) {
+    if (Array.isArray(arg)) arg.push(10); // Should FAIL
+}
+
+function fn7(arg: boolean | number[] | string[], stringAndNumber: string & number) {

```

Comment:
`string & number = never`, I don't think this tests what we want, how about:

```suggestion
function fn7(arg: boolean | Number[] | string[], stringAndNumber: String & Number) {
```